src/query/graphite/native/builtin_functions.go

```diff
@@ -1658,6 +1658,83 @@ func movingMedian(ctx *common.Context, _ singlePathSpec, windowSize string) (*bi
 	}, nil
 }
 
+// movingSum takes one metric or a wildcard seriesList followed by a a quoted string
+// with a length of time like '1hour' or '5min'. Graphs the sum of the preceding
+// datapoints for each point on the graph. All previous datapoints are set to None at
+// the beginning of the graph.
+func movingSum(ctx *common.Context, _ singlePathSpec, windowSize string) (*binaryContextShifter, error) {
+	interval, err := common.ParseInterval(windowSize)
+	if err != nil {
+		return nil, err
+	}
+
+	if interval <= 0 {
+		return nil, common.ErrInvalidIntervalFormat
+	}
+
+	contextShiftingFn := func(c *common.Context) *common.Context {
+		opts := common.NewChildContextOptions()
+		opts.AdjustTimeRange(0, 0, interval, 0)
+		childCtx := c.NewChildContext(opts)
+		return childCtx
+	}
+
+	bootstrapStartTime, bootstrapEndTime := ctx.StartTime.Add(-interval), ctx.StartTime
+	transformerFn := func(bootstrapped, original ts.SeriesList) (ts.SeriesList, error) {
+		bootstrapList, err := combineBootstrapWithOriginal(ctx,
+			bootstrapStartTime, bootstrapEndTime,
+			bootstrapped, singlePathSpec(original))
+		if err != nil {
+			return ts.NewSeriesList(), err
+		}
+
+		results := make([]*ts.Series, 0, original.Len())
+
+		for i, bootstrap := range bootstrapList.Values {
+			series := original.Values[i]
+			windowPoints := int(interval / (time.Duration(series.MillisPerStep()) * time.Millisecond))
+			if windowPoints <= 0 {
+				err := errors.NewInvalidParamsError(fmt.Errorf(
+					"non positive window points, windowSize=%s, stepSize=%d",
+					windowSize, series.MillisPerStep()))
+				return ts.NewSeriesList(), err
+			}
+			window := make([]float64, windowPoints)
+			util.Memset(window, math.NaN())
+			numSteps := series.Len()
+			offset := bootstrap.Len() - numSteps
+			vals := ts.NewValues(ctx, series.MillisPerStep(), numSteps)
+			for i := 0; i < numSteps; i++ {
+				for j := i + offset - windowPoints; j < i+offset; j++ {
+					if j < 0 || j >= bootstrap.Len() {
+						continue
+					}
+
+					idx := j - i - offset + windowPoints
+					if idx < 0 || idx > len(window)-1 {
+						continue
+					}
+
+					window[idx] = bootstrap.ValueAt(j)
+				}
+				vals.SetValueAt(i, common.SafeSum(window))
+			}
+			name := fmt.Sprintf("movingSum(%s,%q)", series.Name(), windowSize)
+			newSeries := ts.NewSeries(ctx, name, series.StartTime(), vals)
+			results = append(results, newSeries)
+		}
+
+		original.Values = results
+		return original, nil
+	}
+
+	return &binaryContextShifter{
+		ContextShiftFunc:  contextShiftingFn,
+		BinaryTransformer: transformerFn,
```

Comment:
Now that I've read the common code, I think the only things that differ is stuff like this:
```go
// movingMedian
				nans := common.SafeSort(window)
				if nans < windowPoints {
					index := (windowPoints - nans) / 2
					median := window[nans+index]
					vals.SetValueAt(i, median)
				}
```

```go
// movingSum
				vals.SetValueAt(i, common.SafeSum(window))
```

And the naming of the series:
```go
// movingAverage
name := fmt.Sprintf("movingSum(%s,%q)", series.Name(), windowSize)
// movingSum
name := fmt.Sprintf("movingSum(%s,%q)", series.Name(), windowSize)
```

Can we extract all the common into a new method? i.e. 
```go
type movingImplFn func(window float64, values ts.MutableValues)

func newMovingBinaryTransform(movingFunctionName string, impl movingImplFn) *binaryContextShifter {
	interval, err := common.ParseInterval(windowSize)
	if err != nil {
		return nil, err
	}

	if interval <= 0 {
		return nil, common.ErrInvalidIntervalFormat
	}

	contextShiftingFn := func(c *common.Context) *common.Context {
		opts := common.NewChildContextOptions()
		opts.AdjustTimeRange(0, 0, interval, 0)
		childCtx := c.NewChildContext(opts)
		return childCtx
	}

	bootstrapStartTime, bootstrapEndTime := ctx.StartTime.Add(-interval), ctx.StartTime
	return &binaryContextShifter{
		ContextShiftFunc:  contextShiftingFn,
		BinaryTransformer: func(bootstrapped, original ts.SeriesList) (ts.SeriesList, error) {
			bootstrapList, err := combineBootstrapWithOriginal(ctx,
				bootstrapStartTime, bootstrapEndTime,
				bootstrapped, singlePathSpec(original))
			if err != nil {
				return ts.NewSeriesList(), err
			}

			results := make([]*ts.Series, 0, original.Len())
			maxWindowPoints := 0
			for i, bootstrap := range bootstrapList.Values {
				series := original.Values[i]
				windowPoints := windowPointsLength(series, interval)
				if windowPoints <= 0 {
					err := errors.NewInvalidParamsError(fmt.Errorf(
						"non positive window points, windowSize=%s, stepSize=%d",
						windowSize, series.MillisPerStep()))
					return ts.NewSeriesList(), err
				}
				if windowPoints > maxWindowPoints {
					maxWindowPoints = windowPoints
				}
			}
			
			windowPoints := make([]float64, maxWindowPoints)
			for i, bootstrap := range bootstrapList.Values {
				series := original.Values[i]
				currWindowPoints := windowPointsLength(series, interval)
				window := windowPoints[:currWindowPoints]
				util.Memset(window, math.NaN())
				numSteps := series.Len()
				offset := bootstrap.Len() - numSteps
				vals := ts.NewValues(ctx, series.MillisPerStep(), numSteps)
				for i := 0; i < numSteps; i++ {
					for j := i + offset - windowPoints; j < i+offset; j++ {
						if j < 0 || j >= bootstrap.Len() {
							continue
						}

						idx := j - i - offset + windowPoints
						if idx < 0 || idx > len(window)-1 {
							continue
						}

						window[idx] = bootstrap.ValueAt(j)
					}
					impl(window, vals)
				}
				name := fmt.Sprintf("%s(%s,%q)", movingFunctionName, series.Name(), windowSize)
				newSeries := ts.NewSeries(ctx, name, series.StartTime(), vals)
				results = append(results, newSeries)
			}

			original.Values = results
			return original, nil
		},
	}
}
```

Reply:
I see that movingAverage differs a fair bit, but at least we could reuse between movingMedian and movingSum, etc?

Reply:
done, great idea!